From f86d165bfe5f6248743774bb07af0bd7cff12443 Mon Sep 17 00:00:00 2001
From: Li Zhengyu <lizhengyu3@huawei.com>
Date: Sat, 28 May 2022 16:06:44 +0800
Subject: [PATCH] arm64: Add non nmi ipi backtrace support
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: f86d165bfe5f6248743774bb07af0bd7cff12443
Modified-by-SEL: No


hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I59IQS
CVE: NA

--------------------------------

Use non nmi ipi to support backtrace on arm64 with nmi
unsupported.
It has been tested on qemu.

Signed-off-by: Li Zhengyu <lizhengyu3@huawei.com>
Reviewed-by: Liao Chang <liaochang1@huawei.com>
Reviewed-by: Wei Li <liwei391@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/arm64/kernel/ipi_nmi.c | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/arch/arm64/kernel/ipi_nmi.c b/arch/arm64/kernel/ipi_nmi.c
index 3b105852fc17..2cf28e511b23 100644
--- a/arch/arm64/kernel/ipi_nmi.c
+++ b/arch/arm64/kernel/ipi_nmi.c
@@ -33,12 +33,24 @@ void arm64_send_nmi(cpumask_t *mask)
 	__ipi_send_mask(ipi_nmi_desc, mask);
 }
 
+static void ipi_cpu_backtrace(void *info)
+{
+	printk_safe_enter();
+	nmi_cpu_backtrace(get_irq_regs());
+	printk_safe_exit();
+}
+
+static void arm64_send_ipi(cpumask_t *mask)
+{
+	smp_call_function_many(mask, ipi_cpu_backtrace, NULL, false);
+}
+
 bool arch_trigger_cpumask_backtrace(const cpumask_t *mask, bool exclude_self)
 {
 	if (!ipi_nmi_desc)
-		return false;
-
-	nmi_trigger_cpumask_backtrace(mask, exclude_self, arm64_send_nmi);
+		nmi_trigger_cpumask_backtrace(mask, exclude_self, arm64_send_ipi);
+	else
+		nmi_trigger_cpumask_backtrace(mask, exclude_self, arm64_send_nmi);
 
 	return true;
 }
-- 
2.33.0

